modules/navier_stokes/src/fvbcs/PINSFVMomentumAdvectionOutflowBC.C

```diff
@@ -0,0 +1,90 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "PINSFVMomentumAdvectionOutflowBC.h"
+#include "PINSFVSuperficialVelocityVariable.h"
+#include "SubProblem.h"
+#include "MooseMesh.h"
+
+registerADMooseObject("NavierStokesApp", PINSFVMomentumAdvectionOutflowBC);
+
+InputParameters
+PINSFVMomentumAdvectionOutflowBC::validParams()
+{
+  InputParameters params = FVMatAdvectionOutflowBC::validParams();
+  params += INSFVFullyDevelopedFlowBC::validParams();
+  params.addRequiredCoupledVar("u", "The velocity in the x direction.");
+  params.addCoupledVar("v", "The velocity in the y direction.");
+  params.addCoupledVar("w", "The velocity in the z direction.");
+  params.addRequiredCoupledVar("porosity", "Porosity auxiliary variable");
+  params.addClassDescription("Outflow boundary condition for advecting momentum. This will impose "
+                             "a zero normal gradient on the boundary velocity.");
+  return params;
+}
+
+PINSFVMomentumAdvectionOutflowBC::PINSFVMomentumAdvectionOutflowBC(const InputParameters & params)
+  : FVMatAdvectionOutflowBC(params),
+    INSFVFullyDevelopedFlowBC(params),
+    _u_var(dynamic_cast<const PINSFVSuperficialVelocityVariable *>(getFieldVar("u", 0))),
+    _v_var(dynamic_cast<const PINSFVSuperficialVelocityVariable *>(getFieldVar("v", 0))),
+    _w_var(dynamic_cast<const PINSFVSuperficialVelocityVariable *>(getFieldVar("w", 0))),
+    _eps(coupledValue("porosity")),
+    _eps_neighbor(coupledNeighborValue("porosity")),
+    _dim(_subproblem.mesh().dimension())
+{
+#ifndef MOOSE_GLOBAL_AD_INDEXING
+  mooseError("PINSFV is not supported by local AD indexing. In order to use PINSFV, please run the "
+             "configure script in the root MOOSE directory with the configure option "
+             "'--with-ad-indexing-type=global'");
+#endif
+
+  if (!_u_var)
+    paramError("u", "the u velocity must be an PINSFVSuperficialVelocityVariable.");
```

Comment:
```suggestion
    paramError("u", "the u velocity must be a PINSFVSuperficialVelocityVariable.");
```
and for other similar